traits/trait_list_object.py

```diff
@@ -53,465 +90,765 @@ def __init__(self, index=0, removed=None, added=None):
+        if index < 0:
+            return max(0, len(self) + index)
+        else:
+            return min(len(self), index)
 
-    def clear(self):
-        if self.trait.minlen > 0:
-            self.len_error(0)
+    def _normalize_slice(self, index):
+        """ Normalize slice start and stop to range 0 to len (inclusive). """
 
-        if len(self) == 0:
-            list.clear(self)
-            return
+        # Do not normalize if step is negative.
+        if index.step is not None and index.step < 0:
+            return index
 
-        removed = self.copy()
-        list.clear(self)
+        if index.step is None or index.step > 0:
+            if index.start is not None:
+                start = self._normalize_index(index.start)
+            else:
+                start = 0
+            if index.stop is not None:
+                stop = self._normalize_index(index.stop)
+            else:
+                stop = len(self)
+        else:
+            if index.start is not None:
+                start = self._normalize_index(index.start)
+            else:
+                start = len(self)
+            if index.stop is not None:
+                stop = self._normalize_index(index.stop)
+            else:
+                stop = 0
 
-        if self.name_items is not None:
-            self._send_trait_items_event(
-                self.name_items, TraitListEvent(0, removed)
-            )
+        return slice(start, stop, index.step)
+
+
+class TraitListObject(TraitList):
+    """ A specialization of TraitList with a default validator and notifier
+    which provide bug-for-bug compatibility with the TraitsListObject from
+    Traits versions before 6.0.
 
-    def len_error(self, len):
-        """ Utility method that raises an error if length is incorrect. """
-        raise TraitError(
-            "The '%s' trait of %s instance must be %s, "
-            "but you attempted to change its length to %d element%s."
-            % (
-                self.name,
-                class_of(self.object()),
-                self.trait.full_info(self.object(), self.name, Undefined),
-                len,
-                "s"[len == 1:],
+    Parameters
+    ----------
+    trait : CTrait
+        The trait that the list has been assigned to.
+    object : HasTraits
+        The object the list belongs to.
+    name : str
+        The name of the trait on the object.
+    value : iterable
+        The initial value of the list.
+    notifiers : list
+        Additional notifiers for the list.
+
+    Attributes
+    ----------
+    trait : CTrait
+        The trait that the list has been assigned to.
+    object : HasTraits
+        The object the list belongs to.
+    name : str
+        The name of the trait on the object.
+    value : iterable
+        The initial value of the list.
+    notifiers : list
+        Additional notifiers for the list.
+    """
+
+    def __init__(self, trait, object, name, value, notifiers=[]):
```

Comment:
The `notifiers` argument here seems to be completely untested. Do we need it? How about we simply remove it? @kitchoi Is this argument something you need for the EEP 3 work?

Reply:
No EEP-3 does not need this in the `__init__`. I agree this can be removed.

Reply:
Removed `notifiers` parameter.